clients/git/client.go

```diff
@@ -0,0 +1,209 @@
+package git
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	cp "github.com/otiai10/copy"
+
+	"github.com/ossf/scorecard/v4/clients"
+)
+
+const repoDir = "repo*"
+
+var (
+	errNilCommitFound = errors.New("nil commit found")
+	errEmptyQuery     = errors.New("query is empty")
+)
+
+type Client struct {
+	gitRepo        *git.Repository
+	worktree       *git.Worktree
+	listCommits    *sync.Once
+	tempDir        string
+	errListCommits error
+	commits        []clients.Commit
+	commitDepth    int
+}
+
+func (c *Client) InitRepo(uri, commitSHA string, commitDepth int) error {
+	// cleanup previous state, if any.
+	c.Close()
+	c.listCommits = new(sync.Once)
+	c.commits = nil
+
+	// init
+	c.commitDepth = commitDepth
+	tempDir, err := os.MkdirTemp("", repoDir)
+	if err != nil {
+		return fmt.Errorf("os.MkdirTemp: %w", err)
+	}
+
+	// git clone
+	const filePrefix = "file://"
+	if strings.HasPrefix(uri, filePrefix) {
+		if err := cp.Copy(strings.TrimPrefix(uri, filePrefix), tempDir); err != nil {
+			return fmt.Errorf("cp.Copy: %w", err)
+		}
+		c.gitRepo, err = git.PlainOpen(tempDir)
+		if err != nil {
+			return fmt.Errorf("git.PlainOpen: %w", err)
+		}
+	} else {
+		c.gitRepo, err = git.PlainClone(tempDir, false /*isBare*/, &git.CloneOptions{
+			URL:      uri,
+			Progress: os.Stdout,
+		})
+		if err != nil {
+			return fmt.Errorf("git.PlainClone: %w", err)
+		}
+	}
+	c.tempDir = tempDir
+	c.worktree, err = c.gitRepo.Worktree()
+	if err != nil {
+		return fmt.Errorf("git.Worktree: %w", err)
+	}
+
+	// git checkout
+	if commitSHA != clients.HeadSHA {
+		if err := c.worktree.Checkout(&git.CheckoutOptions{
+			Hash:  plumbing.NewHash(commitSHA),
+			Force: true, // throw away any unsaved changes.
+		}); err != nil {
+			return fmt.Errorf("git.Worktree: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Client) ListCommits() ([]clients.Commit, error) {
+	c.listCommits.Do(func() {
+		commitIter, err := c.gitRepo.Log(&git.LogOptions{
+			Order: git.LogOrderCommitterTime,
+		})
+		if err != nil {
+			c.errListCommits = fmt.Errorf("git.CommitObjects: %w", err)
+			return
+		}
+		c.commits = make([]clients.Commit, 0, c.commitDepth)
+		for i := 0; i < c.commitDepth; i++ {
+			commit, err := commitIter.Next()
+			if err != nil && !errors.Is(err, io.EOF) {
+				c.errListCommits = fmt.Errorf("commitIter.Next: %w", err)
+				return
+			}
+			// No more commits.
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			if commit == nil {
+				// Not sure in what case a nil commit is returned. Fail explicitly.
+				c.errListCommits = fmt.Errorf("%w", errNilCommitFound)
+				return
+			}
+
+			c.commits = append(c.commits, clients.Commit{
+				SHA:           commit.Hash.String(),
+				Message:       commit.Message,
+				CommittedDate: commit.Committer.When,
+				Committer: clients.User{
+					Login: commit.Committer.Email,
```

Comment:
probably outside the scope of this PR, but changes may be needed here.`DependencyUpdateTool` looks for `Author: "dependabot[bot]"`

Reply:
Yeah, some checks may still not be able to function with `go-git`. We probably need to experiment a bit with `Name` and `Email` fields [here](https://pkg.go.dev/github.com/go-git/go-git/v5@v5.6.0/plumbing/object#Signature) to see if we can find some reasonable approximations.